### ISSUE TITLE
error upgrading gloo from 0.17.6 -> 0.18.0

### ISSUE THREAD

**Issue by rickducott** [solo-io/gloo#902]
- download 0.17.6 CLI
- glooctl install gateway
- deploy petstore demo (https://gloo.solo.io/user_guides/gateway/getting_started/)
- manually update the gateways with some custom HCM settings, i.e. idleTimeout = 10s
- download 0.8.0 CLI 
- glooctl install gateway --upgrade
- gloo gets a fatal error

```
{"level":"fatal","ts":"2019-07-26T12:57:20.042Z","logger":"gloo","caller":"setuputils/main_setup.go:71","msg":"error in setup: starting snapshot watch: initial Proxy list: converting output crd: reading crd spec into v1Proxy: unknown field \"listenerPlugins\" in v1.HttpListener","stacktrace":"github.com/solo-io/gloo/pkg/utils/setuputils.Main\n\t/workspace/gopath/src/github.com/solo-io/gloo/pkg/utils/setuputils/main_setup.go:71\ngithub.com/solo-io/gloo/projects/gloo/pkg/setup.Main\n\t/workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/pkg/setup/setup.go:11\nmain.main\n\t/workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/cmd/main.go:19\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:200"}
```

**Comment by mitchdraft** [comment]
```
0: level: fatal
ts: 2019-07-26T12:57:20.042Z
logger: gloo
caller: setuputils/main_setup.go:71
msg: error in setup: starting snapshot watch: initial Proxy list: converting output crd: reading crd spec into v1Proxy: unknown field "listenerPlugins" in v1.HttpListener
stacktrace: github.com/solo-io/gloo/pkg/utils/setuputils.Main
        /workspace/gopath/src/github.com/solo-io/gloo/pkg/utils/setuputils/main_setup.go:71
github.com/solo-io/gloo/projects/gloo/pkg/setup.Main
        /workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/pkg/setup/setup.go:11
main.main
        /workspace/gopath/src/github.com/solo-io/gloo/projects/gloo/cmd/main.go:19
runtime.main
        /usr/local/go/src/runtime/proc.go:200
```

**Comment by mitchdraft** [comment]
might be related to rename from `ListenerPlugins` to:
`projects/gateway/api/v1/gateway.proto:    gloo.solo.io.HttpListenerPlugins plugins = 5;`

**Comment by mitchdraft** [comment]
https://github.com/solo-io/gloo/commit/d3db948852558cdadef1f07639876aed1892ebd4#diff-81e88746adffe15d2cef77155b186828R121

**Comment by rickducott** [comment]
In 0.17.6, I upgrade the gateway protos with the following block:
```
    plugins:
      httpConnectionManagerSettings:
        idleTimeout: 10s
```

**Comment by mitchdraft** [comment]
rename in the proxy proto. Maybe we should have 'V2'd the proxy like the gateway

`git diff v0.17.6 v0.18.0 -- projects/gloo/api/v1/proxy.proto`


```diff
diff --git a/projects/gloo/api/v1/proxy.proto b/projects/gloo/api/v1/proxy.proto
index aab0450f..5232997a 100644
--- a/projects/gloo/api/v1/proxy.proto
+++ b/projects/gloo/api/v1/proxy.proto
@@ -15,6 +15,7 @@ import "github.com/solo-io/solo-kit/api/v1/solo-kit.proto";
 import "github.com/solo-io/gloo/projects/gloo/api/v1/ssl.proto";
 import "github.com/solo-io/gloo/projects/gloo/api/v1/subset.proto";
 import "github.com/solo-io/gloo/projects/gloo/api/v1/plugins.proto";
+
 /*
 A Proxy is a container for the entire set of configuration that will to be applied to one or more Proxy instances.
 Proxies can be understood as a set of listeners, represents a different bind address/port where the proxy will listen
@@ -65,15 +66,44 @@ message Listener {
         // The HTTP Listener is currently the only supported listener type.
         // It contains configuration options for Gloo's HTTP-level features including request-based routing
         HttpListener http_listener = 4;
+
+        // The HTTP Listener is currently the only supported listener type.
+        // It contains configuration options for GLoo's HTTP-level features including request-based routing
+        TcpListener tcp_listener = 5;
     }
 
     // SSL Config is optional for the listener. If provided, the listener will serve TLS for connections on this port.
     // Multiple SslConfigs are supported for the purpose of SNI. Be aware that the SNI domain provided in the SSL Config
-    // must match a domain in virtual host.
-    repeated SslConfig ssl_configurations = 5;
+    repeated SslConfig ssl_configurations = 6;
 
     // Enable ProxyProtocol support for this listener
-    google.protobuf.BoolValue use_proxy_proto = 6;
+    google.protobuf.BoolValue use_proxy_proto = 7;
+
+    // top level plugins
+    ListenerPlugins plugins = 8;
+
+}
+
+message TcpListener {
+    // List of filter chains to match on for this listener
+    repeated TcpHost tcp_hosts = 1;
+    // Plugins contains top-level plugin configuration to be applied to a listener
+    // Listener config is applied to traffic for the given listener.
+    // Some configuration here can be overridden in
+    // Virtual Host Plugin configuration or Route Plugin configuration
+    TcpListenerPlugins plugins = 8;
+}
+
+message TcpHost {
+    // the logical name of the tcp host. names must be unique for each tcp host within a listener
+    string name = 1;
+    // Name of the destinations the gateway can route to.
+    // Note: the destination spec and subsets are not supported in this context and will be ignored.
+    RouteAction destination = 2;
+    // If provided, the Gateway will serve TLS/SSL traffic for this set of routes
+    gloo.solo.io.SslConfig ssl_config = 3;
+    //matcher
+
 }
 
 // Use this listener to configure proxy behavior for any HTTP-level features including defining routes (via virtual services).
@@ -84,12 +114,11 @@ message HttpListener {
     // at least one virtual host must be specified for this listener to be active (else connections will be refused)
     // the set of domains for each virtual host must be unique, or the config will be considered invalid
     repeated VirtualHost virtual_hosts = 1;
-
     // Plugins contains top-level plugin configuration to be applied to a listener
-    // Listener config is applied to all HTTP traffic that
-    // connects to this listener. Some configuration here can be overridden in
+    // Listener config is applied to traffic for the given listener.
+    // Some configuration here can be overridden in
     // Virtual Host Plugin configuration or Route Plugin configuration
-    HttpListenerPlugins listener_plugins = 2;
+    HttpListenerPlugins plugins = 2;
 }
 
 /*
@@ -261,6 +290,9 @@ message Destination {
         ConsulServiceDestination consul = 12;
     }
 
+
+    // Note: When in the context of a TCP gateway the below 2 fields will be ignored
+
     // Some upstreams utilize plugins which require or permit additional configuration on routes targeting them.
     // gRPC upstreams, for example, allow specifying REST-style parameters for JSON-to-gRPC transcoding in the
     // destination config. If the destination config is required for the upstream and not provided by the user,
```

**Comment by rickducott** [comment]
No, the rename was intentional. The intention was for gloo v0.18.0 to only configure gateway-proxy-v2, but it is seeing both proxy objects in the snapshot. If you manually delete the old proxy object, then gloo works as expected, the old gateway-proxy pod doesn't get reconfigured, and everything is OK -- however, that is not a desired workaround because we wanted to preserve the ability to rollback, and not require deleting v1 resources. I'd like to look into Gloo filtering out old proxy from the snapshot.

**Comment by rickducott** [comment]
On second thought, doesn't seem like we renamed it for a good reason and we could just do that.

**Comment by rickducott** [comment]
Should be fixed in #903
